proposals/NNNN-swiftpm-package-traits.md

```diff
@@ -0,0 +1,479 @@
+            enabledTraits: ["SomeTrait"],
+            disableDefaultTraits: true
+        )
+    ),
+]
+```
+
+Another common scenario is to enable a trait of a dependency only when a trait
+of the package is enabled. The below example enables the `SomeOtherTrait` when
+the `Foo` trait of this package is enabled.
+
+```swift
+dependencies: [
+    .package(
+        url: "https://github.com/Org/SomePackage.git",
+        from: "1.0.0",
+        from: "1.0.0",
+        traits: Package.Dependency.Trait(
+            enabledTraits: [
+                "SomeTrait",
+                EnabledTrait("SomeOtherTrait", condition: .when(traits: ["Foo"])),
+            ]
+        )
+    ),
+]
+```
+
+Conditional dependencies are specified per target and extend the current
+`condition` syntax which is used for specifying platform dependent dependencies.
+
+```swift
+targets: [
+    .target(
+        name: "SomeTarget",
+        dependencies: [
+            .product(
+                name: "SomeProduct",
+                package: "SomePackage",
+                condition: .when(traits: ["Foo"])
+            ),
+        ]
+    )
+]
+```
+
+Lastly, code can be conditionally compiled by checking if a trait is enabled.
+This can be used for both optional dependencies by surrounding the `import`
+statements in a trait check and for regular code where you want to modify its
+behaviour depending on the enabled traits.
+
+```swift
+#if TRAIT_Foo
+import SomeDependency
+#endif
+
+func hello() {
+    #if TRAIT_Foo
+    Foo.hello()
+    #else
+    print("Hello")
+    #endif
+}
+```
+
+At this point, it is important to talk about the trait unification across the
+entire dependency graph. After dependency resolution the union of enabled traits
+per package is calculated. This is then used to determine both the enabled
+optional dependencies and the enabled traits for the compile time checks. Since
+the enabled traits of a dependency are specified on a per package level and not
+from the root of the tree, any combination of enabled traits must be supported.
+A consequence of this is that all traits **must** be additive. Enabling a trait
+**must never** disable functionality i.e. remove API or lead to any other
+SemVer-incompatible change.
+
+## Detailed design
+
+This proposal extends the current `PackageDescription` APIs by introducing the
+following new `Trait` type.
+
+```swift
+/// A struct representing a package's trait.
+///
+/// Traits can be used for expressing conditional compilation and optional dependencies.
+///
+/// - Important: Traits must be strictly additive and enabling a trait **must not** remove API.
+public struct Trait: Hashable, ExpressibleByStringLiteral {
+    /// The trait's canonical name.
+    ///
+    /// This is used when enabling the trait or when referring to it from other modifiers in the manifest.
+    public var name: String
+
+    /// A set of other traits of this package that this trait enables.
+    public var enabledTraits: Set<String>
```

Comment:
I don't find this naming very obvious when used in the example above. The only suggestions which come to mind are `dependencies` and `subTraits`, neither of which seem quite right.

Reply:
I agree. I was struggling a lot with naming this. Initial I had just made this a `Dictionary<String: String>` but wanted some more strongly typed.

Reply:
I agree that "dependencies" or "dependentTraits" would be clearer, an in line with package and target dependency syntax. Otherwise it's a bit confusing that a trait "has" traits (not sure what that means).